components/datetime/src/format/datetime.rs

```diff
@@ -668,69 +675,66 @@ where
-                    .and_then(|ns| {
-                        // We only support fixed field length for fractional seconds.
-                        let FieldLength::Fixed(p) = length else {
-                            return Err(DateTimeWriteError::UnsupportedField(next_field));
-                        };
-                        Ok((ns, p))
-                    });
-                match r {
-                    Err(e) => {
+            Some(second) => {
+                match match (iter.peek(), formatting_options.fractional_second_digits) {
+                    (
+                        Some(&PatternItem::Field(
+                            next_field @ Field {
+                                symbol: FieldSymbol::Second(Second::FractionalSecond),
+                                length: FieldLength::Fixed(p),
+                            },
+                        )),
+                        _,
+                    ) => {
+                        // Fractional second digits via field symbol
+                        iter.next(); // Advance over nanosecond symbol
+                        Some((-(p as i16), Some(next_field), datetime.nanosecond()))
+                    }
+                    (_, Some(p)) => {
+                        // Fractional second digits via semantic option
+                        Some((-(p as i16), None, datetime.nanosecond()))
+                    }
+                    _ => None,
+                } {
+                    Some((_, maybe_next_field, None)) => {
+                        // Request to format nanoseconds but we don't have nanoseconds
                         let seconds_result =
                             try_write_number(w, fdf, usize::from(second).into(), l)?;
-                        write_value_missing(w, next_field)?;
+                        if let Some(next_field) = maybe_next_field {
+                            write_value_missing(w, next_field)?;
+                        }
                         // Return the earlier error
-                        seconds_result.and(Err(e))
+                        seconds_result.and(Err(DateTimeWriteError::MissingInputField("nanosecond")))
                     }
-                    Ok((ns, p)) => {
+                    Some((position, maybe_next_field, Some(ns))) => {
+                        // Formatting with fractional seconds
                         let mut s = FixedDecimal::from(usize::from(second));
                         let _infallible = s.concatenate_end(
                             FixedDecimal::from(usize::from(ns)).multiplied_pow10(-9),
                         );
                         debug_assert!(_infallible.is_ok());
-                        s.pad_end(-(p as i16));
+                        s.pad_end(position);
+                        if maybe_next_field.is_none() {
+                            // Truncate on semantic option but not "S" field
+                            // TODO: Does this make sense?
```

Comment:
(I don't understand the context here)

Reply:
There are two ways to hit this code path

1. A pattern such as "H:mm:ssSSS"
2. fractionalSecondDigits option

The behavior implemented for "SSS" had been that it is a minimum but not maximum number of digits. 

I'm currently implementing fSD as both minimum and maximum.